### PR TITLE
Standalone UI

### PR DESCRIPTION
**Pull request by 200Tigersbloxed**
## Description 📙

The Standalone UI PR implements #18. Technically still reliant on Counters+ (as per request), however can be bridged away from Counters+ easily. Supports Standard, 90Degree, and 360Degree maps, excluding Multiplayer maps.

## Innerworkings 🤓

The Standalone UI is non-reliant on Beat Saber libraries (excluding [built-in shaders](https://github.com/200Tigersbloxed/HRCounter/blob/standalone/HRCounter/AssetBundleManager.cs#L142-L143)) and mainly uses Reflection for any Beat Saber values, so random exceptions after updates are minimal. Assets are completely based off of Unity's [AssetBundles](https://docs.unity3d.com/Manual/AssetBundlesIntro.html) ([the AssetBundle](https://github.com/200Tigersbloxed/HRCounter/blob/standalone/HRCounter/hrcounter)) and is loaded from memory using EmbeddedResources from the Executing Assembly's ManifestResourceStream. This means that the UI can be made directly in the UnityEngine and all Assets ([excluding scripts](https://forum.unity.com/threads/including-scripts-in-assetbundles.459849/)) can be loaded at runtime with minimal effort.

### + [AssetBundleManager.cs](https://github.com/200Tigersbloxed/HRCounter/blob/ef00f3ad9485b0a69244954b40e163c8fa866db9/HRCounter/AssetBundleManager.cs)

Handles [loading the AssetBundle](https://github.com/200Tigersbloxed/HRCounter/blob/ef00f3ad9485b0a69244954b40e163c8fa866db9/HRCounter/AssetBundleManager.cs#L21-L39), [instantiating it's packaged prefab](https://github.com/200Tigersbloxed/HRCounter/blob/ef00f3ad9485b0a69244954b40e163c8fa866db9/HRCounter/AssetBundleManager.cs#L32), [checking game values to see if it can actually be instanced](https://github.com/200Tigersbloxed/HRCounter/blob/ef00f3ad9485b0a69244954b40e163c8fa866db9/HRCounter/AssetBundleManager.cs#L41-L124), [instancing and destroying the canvas](https://github.com/200Tigersbloxed/HRCounter/blob/ef00f3ad9485b0a69244954b40e163c8fa866db9/HRCounter/AssetBundleManager.cs#L131-L171), [setting the text of the counter](https://github.com/200Tigersbloxed/HRCounter/blob/ef00f3ad9485b0a69244954b40e163c8fa866db9/HRCounter/AssetBundleManager.cs#L173-L182), and [handling position and rotation on 90/360Degree maps](https://github.com/200Tigersbloxed/HRCounter/blob/ef00f3ad9485b0a69244954b40e163c8fa866db9/HRCounter/AssetBundleManager.cs#L184-L205). 

Essentially, the process is get the AssetBundle's bytes from EmbeddedResources, [Load the AssetBundle from memory](https://docs.unity3d.com/ScriptReference/AssetBundle.LoadFromMemory.html), [get the prefab from the AssetBundle](https://docs.unity3d.com/ScriptReference/AssetBundle.LoadAsset.html), [unload the AssetBundle while keeping all the Assets loaded](https://docs.unity3d.com/ScriptReference/AssetBundle.Unload.html), then whenever requested to create, verify that we can create, and if we can, then create a new canvas in the scene, position and rotate it based on whether it's a Standard or 90/360Degree map, then, once requested, delete the Canvas and reset all values.

### - [PluginConfig.cs](https://github.com/200Tigersbloxed/HRCounter/blob/ef00f3ad9485b0a69244954b40e163c8fa866db9/HRCounter/Configuration/PluginConfig.cs)

Added a [new config value](https://github.com/200Tigersbloxed/HRCounter/blob/ef00f3ad9485b0a69244954b40e163c8fa866db9/HRCounter/Configuration/PluginConfig.cs#L41) called `UseCountersPlus` which will fallback on Counters+ if a user requests (as per request). Controllable with the new BSML config.

### - [HRCounter.cs](https://github.com/200Tigersbloxed/HRCounter/blob/ef00f3ad9485b0a69244954b40e163c8fa866db9/HRCounter/HRCounter.cs)

Will now handle [whether the new Standalone UI should be used, or if Counters+ should be used](https://github.com/200Tigersbloxed/HRCounter/blob/ef00f3ad9485b0a69244954b40e163c8fa866db9/HRCounter/HRCounter.cs#L146-L164). Essentially, [the Standalone UI and Counters+ will both be used](https://github.com/200Tigersbloxed/HRCounter/blob/ef00f3ad9485b0a69244954b40e163c8fa866db9/HRCounter/HRCounter.cs#L127-L133), however, [only one will be visible](https://github.com/200Tigersbloxed/HRCounter/blob/ef00f3ad9485b0a69244954b40e163c8fa866db9/HRCounter/HRCounter.cs#L151-L161). Can be changed in real-time, even in the middle of a map!

### - [HRCounter.csproj](https://github.com/200Tigersbloxed/HRCounter/blob/standalone/HRCounter/HRCounter.csproj)

Changed `Newtonsoft.Json` to use the dll from Beat Saber's Managed folder. This can be changed back if preferred. Bumped small libraries' versions to a later version (0Harmony and BSML), however it isn't necessary to keep, nor does it make a difference. Added `UnityEngine.AssetBundleModule` as a dependency; this library is required.

### - [Plugin.cs](https://github.com/200Tigersbloxed/HRCounter/blob/ef00f3ad9485b0a69244954b40e163c8fa866db9/HRCounter/Plugin.cs)

Loads the AssetBundle when [Init(IPALogger)](https://github.com/200Tigersbloxed/HRCounter/blob/ef00f3ad9485b0a69244954b40e163c8fa866db9/HRCounter/Plugin.cs#L34) is invoked. This should be done regardless of if a user wants to use the Standalone UI or not, when checking on Init.

### - [UI/BSML/configMenu.bsml](https://github.com/200Tigersbloxed/HRCounter/blob/ef00f3ad9485b0a69244954b40e163c8fa866db9/HRCounter/UI/BSML/configMenu.bsml)

Added the [`UseCountersPlus`](https://github.com/200Tigersbloxed/HRCounter/blob/ef00f3ad9485b0a69244954b40e163c8fa866db9/HRCounter/UI/BSML/configMenu.bsml#L15) bool-setting.

### - [UI/SettingMenuController.cs](https://github.com/200Tigersbloxed/HRCounter/blob/ef00f3ad9485b0a69244954b40e163c8fa866db9/HRCounter/UI/SettingMenuController.cs)

Registered the [`UseCountersPlus`](https://github.com/200Tigersbloxed/HRCounter/blob/ef00f3ad9485b0a69244954b40e163c8fa866db9/HRCounter/UI/SettingMenuController.cs#L99-L107) UIValue.

### - [Utils/GamePause.cs](https://github.com/200Tigersbloxed/HRCounter/blob/ef00f3ad9485b0a69244954b40e163c8fa866db9/HRCounter/Utils/GamePause.cs)

Handles [removing](https://github.com/200Tigersbloxed/HRCounter/blob/ef00f3ad9485b0a69244954b40e163c8fa866db9/HRCounter/Utils/GamePause.cs#L18-L19) the AssetBundle when returning to the Main Menu. Was unsure if the [HRCounter](https://github.com/200Tigersbloxed/HRCounter/blob/standalone/HRCounter/HRCounter.cs#L13) base class had an override for this, however this seems to work.

### + [hrcounter](https://github.com/200Tigersbloxed/HRCounter/blob/ef00f3ad9485b0a69244954b40e163c8fa866db9/HRCounter/hrcounter)

The AssetBundle that contains all the Standalone UI's Assets. If you wish to see it's contents, use [AssetBundles-Browser](https://github.com/Unity-Technologies/AssetBundles-Browser) (also integrated into Unity's Package Manager). Built with Unity version 2019.4.31f1

## Final Notes 📝

I did not change the versioning in [AssemblyInfo.cs](https://github.com/200Tigersbloxed/HRCounter/blob/standalone/HRCounter/Properties/AssemblyInfo.cs), so that's up to you what version you'd like it to be. Happy reviewing!

### CODE REVIEW

File: HRCounter/HRCounter.csproj [hunk 1]
```diff
@@ -42,10 +42,10 @@
     <DisableZipRelease>True</DisableZipRelease>
   </PropertyGroup>
   <ItemGroup>
-    <Reference Include="0Harmony, Version=2.0.2.0, Culture=neutral, PublicKeyToken=null">
+    <Reference Include="0Harmony, Version=2.7.0.0, Culture=neutral, PublicKeyToken=null">
       <HintPath>$(BeatSaberDir)\Libs\0Harmony.dll</HintPath>
     </Reference>
-    <Reference Include="BSML, Version=1.5.2.0, Culture=neutral, processorArchitecture=MSIL">
+    <Reference Include="BSML, Version=1.6.1.0, Culture=neutral, processorArchitecture=MSIL">
```
Comment (qe201020335) [inline HRCounter/HRCounter.csproj:48]: Please revert the changes to the version of the dependencies since I want this mod to work on the old versions of games.

File: HRCounter/HRCounter.csproj [hunk 2]
```diff
@@ -55,9 +55,9 @@
       <HintPath>$(BeatSaberDir)\Plugins\Counters+.dll</HintPath>
       <SpecificVersion>False</SpecificVersion>
     </Reference>
-    <Reference Include="Newtonsoft.Json, Version=12.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed, processorArchitecture=MSIL">
+    <Reference Include="Newtonsoft.Json, Version=8.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed, processorArchitecture=MSIL">
       <Private>False</Private>
-      <HintPath>$(BeatSaberDir)\Libs\Newtonsoft.Json.dll</HintPath>
+      <HintPath>$(BeatSaberDir)\Beat Saber_Data\Managed\Newtonsoft.Json.dll</HintPath>
```
Comment (qe201020335) [inline HRCounter/HRCounter.csproj:60]: These two changes are not necessary anymore because the new BSIPA will force the 12.0 version to load first i.e the one in the `Libs` folder

**Review by qe201020335 — CHANGES_REQUESTED** [review]

**Comment by qe201020335** [comment]
One side note,
You should add a `HRCounter.csproj.user` file beside `HRCounter.csproj` with the contents similar to 
```xml
<?xml version="1.0" encoding="utf-8"?>
<Project ToolsVersion="Current" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <PropertyGroup>
    <BeatSaberDir>U:\SteamLibrary\steamapps\common\Beat Saber</BeatSaberDir>
    <ReferencePath>U:\SteamLibrary\steamapps\common\Beat Saber\Beat Saber_Data\Managed;U:\SteamLibrary\steamapps\common\Beat Saber\Libs;U:\SteamLibrary\steamapps\common\Beat Saber\Plugins</ReferencePath>
  </PropertyGroup>
</Project>
```
So that you don't have to edit `HRCounter.csproj` for the reference paths

File: HRCounter/HRCounter.csproj (hunk 1 above)
Comment (200Tigersbloxed) [reply HRCounter/HRCounter.csproj]: Reverted BSML changes in csproj

File: HRCounter/HRCounter.csproj (hunk 2 above)
Comment (200Tigersbloxed) [reply HRCounter/HRCounter.csproj]: Reverted Newtonsoft.Json changes in csproj

**Review by qe201020335 — APPROVED** [review]
